core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java

```diff
@@ -354,69 +349,27 @@ public EdgeCache(CacheConfiguration conf) {
+    public Map<String, Long> getChildIds(Long inodeId) {
       if (mBackingStoreEmpty) {
-        return mIdToChildMap.getOrDefault(inodeId, mEmpty);
+        return mIdToChildMap.getOrDefault(inodeId, Collections.emptyMap());
       }
       // This implementation must be careful because edges can be asynchronously evicted from the
-      // cache to the backing store.
-      //
-      // To prevent races, we copy the cached children and unflushed deletes *before* inspecting the
-      // backing store. This fulfills the first consistency guarantee because pre-existing children
-      // move in the direction of cache -> backing store, so by checking in this order we are
-      // guaranteed to find them. The mUnflushedDeletes map prevents inclusion of inodes which have
-      // been removed from the cache, but not yet flushed to the backing store. This could happen
-      // if we create, evict, load, then remove an edge. Before the remove is evicted, the edge will
-      // be returned by the backing store's getChildIds. mUnflushedDeletes filters out such phantom
-      // edges.
-      Iterator<Map.Entry<String, Long>> cachedChildren =
-          new ArrayList<>(mIdToChildMap.getOrDefault(inodeId, mEmpty).entrySet()).iterator();
+      // cache to the backing store. To account for this, we read from the cache before consulting
+      // the backing store.
+      Map<String, Long> childIds = new HashMap<>();
+      mIdToChildMap.getOrDefault(inodeId, Collections.emptyMap()).forEach((name, id) -> {
+        childIds.put(name, id);
+      });
+      // Copy the list of unflushed deletes before reading the backing store to prevent racing async
+      // deletion.
       Set<String> unflushedDeletes =
           new HashSet<>(mUnflushedDeletes.getOrDefault(inodeId, Collections.EMPTY_SET));
-      Iterator<Long> flushedChildren = mBackingStore.getChildIds(inodeId).iterator();
-      ConcurrentSkipListMap<String, Long> result = new ConcurrentSkipListMap<>();
-
-      Map.Entry<String, Long> cached = nextCachedInode(cachedChildren);
-      Inode flushed = nextFlushedInode(flushedChildren, unflushedDeletes);
-      while (cached != null && flushed != null) {
-        int comparison = cached.getKey().compareTo(flushed.getName());
-        if (comparison == 0) {
-          // De-duplicate children with the same name.
-          result.put(cached.getKey(), cached.getValue());
-          flushed = nextFlushedInode(flushedChildren, unflushedDeletes);
-          cached = nextCachedInode(cachedChildren);
-        } else if (comparison < 0) {
-          result.put(cached.getKey(), cached.getValue());
-          cached = nextCachedInode(cachedChildren);
-        } else {
-          result.put(flushed.getName(), flushed.getId());
-          flushed = nextFlushedInode(flushedChildren, unflushedDeletes);
+      mBackingStore.getChildren(inodeId).forEach(inode -> {
+        if (!unflushedDeletes.contains(inode.getName())) {
```

Comment:
could two inode with the same name cause problem here?

Reply:
This is in the context of listing children of a single directory, so the only way there can be multiple inodes with the same name is if inodes are concurrently added/deleted/renamed during the call to `getChildIds`. In that case, we make no guarantees about whether the concurrently modified children are included in the listing

Reply:
Got it.